src/gifwrite.c

```diff
@@ -12,15 +12,58 @@ void generate_GIF_data (struct context * context) {
   if ((uint8_t) (depth >> 8) > overall) overall = depth >> 8;
   if ((uint8_t) (depth >> 16) > overall) overall = depth >> 16;
   if (overall > 8) overall = 8;
-  header[4] = (overall - 1) << 4;
-  header[5] = header[6] = 0;
+  bytewrite(header + 4, (overall - 1) << 4, 0, 0);
   if (context -> source -> palette)
     generate_GIF_data_with_palette(context, header);
   else
     generate_GIF_data_from_raw(context, header);
   byteoutput(context, 0x3b);
 }
 
+#define checkboundaries(buffertype, buffer, frame) do {                                                                          \
```

Comment:
The `generate_GIF_data_with_palette` and `generate_GIF_frame_data` cases were very similar, so I factored them out into this macro.